pkg/server/server.go

```diff
@@ -15,25 +15,41 @@ import (
 	"github.com/spf13/viper"
 	"github.com/ziflex/lecho/v3"
 
+	"github.com/zapier/kubechecks/pkg/app_watcher"
 	"github.com/zapier/kubechecks/pkg/config"
 	"github.com/zapier/kubechecks/pkg/vcs"
 )
 
 const KubeChecksHooksPathPrefix = "/hooks"
 
 type Server struct {
-	cfg *config.ServerConfig
+	cfg        *config.ServerConfig
+	appWatcher *app_watcher.ApplicationWatcher
 }
 
 func NewServer(cfg *config.ServerConfig) *Server {
-	return &Server{cfg: cfg}
+	var appWatcher *app_watcher.ApplicationWatcher
+	if viper.GetBool("monitor-all-applications") {
+		argoMap, err := config.BuildAppsMap(context.TODO())
+		if err != nil {
+			log.Fatal().Err(err).Msg("could not build VcsToArgoMap")
+		}
+		cfg.VcsToArgoMap = argoMap
+
+		appWatcher, err = app_watcher.NewApplicationWatcher(cfg)
+		if err != nil {
+			log.Fatal().Err(err).Msg("could not create ApplicationWatcher")
+		}
+	} else {
+		cfg.VcsToArgoMap = config.NewVcsToArgoMap()
+	}
+
+	return &Server{cfg: cfg, appWatcher: appWatcher}
 }
 
 func (s *Server) Start(ctx context.Context) {
-	if argoMap, err := s.buildVcsToArgoMap(ctx); err != nil {
-		log.Warn().Err(err).Msg("failed to build vcs app map from argo")
-	} else {
-		s.cfg.VcsToArgoMap = argoMap
+	if s.appWatcher != nil {
+		go s.appWatcher.Run(context.Background(), 1)
```

Comment:
I _think_ we want this to be `ctx`, not a new context, so it gets canceled when the app gets canceled